clang/lib/Basic/Targets/AArch64.cpp

```diff
@@ -794,14 +795,22 @@ AArch64TargetInfo::getTargetBuiltins() const {
 
 std::optional<std::pair<unsigned, unsigned>>
 AArch64TargetInfo::getVScaleRange(const LangOptions &LangOpts,
-                                  bool IsArmStreamingFunction) const {
+                                  bool IsArmStreamingFunction,
+                                  llvm::Function *F) const {
   if (LangOpts.VScaleMin || LangOpts.VScaleMax)
     return std::pair<unsigned, unsigned>(
         LangOpts.VScaleMin ? LangOpts.VScaleMin : 1, LangOpts.VScaleMax);
 
   if (hasFeature("sve") || (IsArmStreamingFunction && hasFeature("sme")))
     return std::pair<unsigned, unsigned>(1, 16);
 
+  if (F && F->hasFnAttribute("target-features")) {
+    StringRef Str = F->getFnAttribute("target-features").getValueAsString();
+    for (const auto &s : llvm::split(Str, ",")) {
+      if (s == "+sve")
```

Comment:
Should we also be doing something similar for sme in streaming mode, i.e.

```
  if (s == "+sve" || (IsArmStreamingFunction && s == "+sme"))
```

? Might also be worth having a test for this if you agree it makes sense.

Reply:
I agree it makes sense, but the reason I hesitated was IsArmStreamingFunction looks like it could become an internal function of FunctionDecl which should help keep CodeGenFunction.cpp target agnostic. I thought it might be better to submit the changes to SME streaming mode as a separate patch.

Reply:
Fair enough!